### PR TITLE
Enhance BackendMemory to provide more allocation options

### PR DESCRIPTION
**Pull request by deadeyegoodwin**

### CODE REVIEW

**Comment by CoderHam** [comment]
Since TRITONSERVER_MemoryType can be derived from AllocationType it is redundant to pass both to the BackendMemory constructor.
Instead we can have a AllocationTypeToMemoryType() that maintains the mapping for the same and is called when needed.

File: src/backend_memory.cc [hunk 1]
```diff
@@ -26,53 +26,141 @@
...
-  TRITONSERVER_Error* err = TRITONBACKEND_MemoryManagerAllocate(
-      manager, &ptr, memory_type, memory_type_id, byte_size);
-  if (err != nullptr) {
-    TRITONSERVER_ErrorDelete(err);
-
-    // Fallback to CPU memory...
-    memory_type = TRITONSERVER_MEMORY_CPU;
-    RETURN_IF_ERROR(TRITONBACKEND_MemoryManagerAllocate(
-        manager, &ptr, memory_type, 0 /* memory_type_id */, byte_size));
+  bool found = false;
```
Comment (CoderHam) [inline src/backend_memory.cc:136]: found -> success

File: src/backend_memory.cc [hunk 2]
```diff
@@ -26,53 +26,141 @@
...
+    case AllocationType::CPU_PINNED_POOL: {
+      memory_type = TRITONSERVER_MEMORY_CPU_PINNED;
+      RETURN_IF_ERROR(TRITONBACKEND_MemoryManagerAllocate(
+          manager, &ptr, TRITONSERVER_MEMORY_CPU_PINNED, memory_type_id,
+          byte_size));
+      break;
+    }
+
+    case AllocationType::GPU_POOL: {
+      memory_type = TRITONSERVER_MEMORY_GPU;
```
Comment (CoderHam) [inline src/backend_memory.cc:110]: wrap in ifdef/else/endif TRITON_ENABLE_GPU

**Review by CoderHam — CHANGES_REQUESTED** [review]

File: src/backend_memory.cc (hunk 1 above)
Comment (deadeyegoodwin) [reply src/backend_memory.cc:136]: done

File: src/backend_memory.cc (hunk 2 above)
Comment (deadeyegoodwin) [reply src/backend_memory.cc:110]: This is taken care of inside of TRITONBACKEND_MemoryManagerAllocate

**Review by CoderHam — APPROVED** [review]
Looks good, just waiting for the CI to pass. Not sure if the current one picked up the latest commit for correctly freeing allocations

**Pull request merged**